Upgrade to Kubernetes 1.8 API

chazsconi opened the pull request:
This upgrades the API to use K8S 1.8

I have kept the namespaces as they were previously in Kazan so this should be mostly backwards compatible. 

The `model_test.exs` is changed as `Kazan.Models.Apis.Extensions.V1beta1.APIVersion` no longer exists in 1.8, so instead another model in this namespace is used.

chazsconi commented on test/models_test.exs:
```diff
@@ -41,16 +41,16 @@ defmodule KazanModelsTest do
     test "that we can decode a model that contains a v1beta1 kind" do
       {:ok, result} = Models.decode(
         %{
-          "kind" => "APIVersion",
```
APIVersion is not available in 1.8

chazsconi commented:
The tests are failing on Elixir 1.4.4/OTP 19.1 because they just take too long to run, presumably in generating the code from the Swagger file.  Not sure if something can be done about this?

obmarg commented:
> The tests are failing on Elixir 1.4.4/OTP 19.1 because they just take too long to run, presumably in generating the code from the Swagger file. Not sure if something can be done about this?

Yeah, that's odd.  I've also seen a crash on one of my builds using that combination.  It is odd, but I don't have much time to investigate.  May be easiest to just drop support for that combination - if anyone particularly needs it they're welcome to look into fixing it.

obmarg commented:
> I have kept the namespaces as they were previously in Kazan so this should be mostly backwards compatible.

I'm a bit unsure about this.  Backwards compatability is good, but kazan is pre-1.0.0 and the k8s specs themselves seem to be in a bit of a state of flux.  I'm wondering if it'd make sense to switch to something closer to how they're grouped in k8s now:

- `io.k8s.api.*` -> `Kazan.Models.*`.
- `io.k8s.apimachinery.pkg.*" -> `Kazan.Models.ApiMachinery`
- `io.k8s.*` -> `Kazan.Models.*`

The current way of doing things where we've got some stuff under `Kazan.Models.Api` and some under `Kazan.Models.Apis.*` made sense with the old k8s specs, but seems messier now.

chazsconi commented:
So the `Api`/`Apis`/`pkg` parts of the namespace should be removed completely?

In this case I think this mapping should also be added:
`io.k8s.kube-aggregator.pkg.apis.` -> `Kazan.Models.KubeAggregator` otherwise `pkg.apis` will be left in this case.

Also, should the namespaces be converted to camel case just using normal rules - e.g. first letter, and after a `-` is in upper case?

e.g.`apiextensions-apiserver` maps to `ApiextensionsApiserver` rather than `ApiExtensionsApiServer`.

This is how it is done in the Go API client.  https://godoc.org/k8s.io/client-go/kubernetes

obmarg commented:
> So the Api/Apis/pkg parts of the namespace should be removed completely?

I kinda think so.  We've already got `Kazan.Api` - having 95% of the models be under `Kazan.Models.Api` seems a bit verbose.

> In this case I think this mapping should also be added:
io.k8s.kube-aggregator.pkg.apis. -> Kazan.Models.KubeAggregator otherwise pkg.apis will be left in this case.

Yep.  I didn't list this, but I agree.

> Also, should the namespaces be converted to camel case just using normal rules - e.g. first letter, and after a - is in upper case?

I'm not sure about this.  I do like that you've changed `autoscaling` to `AutoScaling` in the PR.  However, I would like to keep things consistent between APIs and Models, so if we do the conversion in one place I would like to do it in the other as well....

obmarg commented:
Another (more extreme) option that occurred to me last night: maybe the `Kazan.Models.Api` module doesn't actually need to exist.  Most of the modules under there correspond with one of the modules under `Kazan.Api` - so _maybe_ they'd be better off living there.

So `Kazan.Models.Apis.Apps.V1.*` would move into `Kazan.Apis.AppsV1` or similar...

I'm not sure exactly how feasible this is, just putting the idea out there...

chazsconi commented:
I think it's best to keep the automated camel-casing, then the names will be consistent everywhere,  the mapping is simpler, and otherwise a mapping would be needed for the Apis also.

Not sure how to make the changes for the more "extreme" renaming you describe.  Perhaps this should be done as a separate release.

So I now have this for mapping:
```
  defp strip_module_name_prefixes(name) do
    case name do
      # Deprecated
      "io.k8s.kubernetes.pkg.api." <> rest ->
        "api." <> rest
      # Deprecated
      "io.k8s.kubernetes.pkg.apis." <> rest ->
        "apis." <> rest

      "io.k8s.api.core." <> rest ->
        rest
      "io.k8s.api." <> rest ->
        rest
      "io.k8s.apimachinery.pkg.apis." <> rest ->
        "Apimachinery." <> rest
      "io.k8s.apimachinery.pkg." <> rest ->
        "Apimachinery." <> rest
      "io.k8s.kube-aggregator.pkg.apis." <> rest ->
        "KubeAggregator." <> rest
      "io.k8s.apiextensions-apiserver.pkg.apis." <> rest ->
        "ApiextensionsApiserver." <> rest
    end
  end
```

If this is OK, I'll fix the tests to match.

obmarg commented:
> I think it's best to keep the automated camel-casing, then  the names will be consistent everywhere,  the mapping is simpler, and  otherwise a mapping would be needed for the Apis also.

Cool - I may end up reintroducing the proper camel casing at some point, but happy to stick with the automatable solution for this PR.

> Not sure how to make the changes for the more "extreme" renaming you describe. Perhaps this should be done as a separate release.

Yeah, that's fair, I was just thinking aloud really.  Might end up doing this, but happy to leave it as is for this PR.

The code sample you've provided looks good, with one minor corrections: could we get rid of the 

```
      "io.k8s.api.core." <> rest ->
        rest
```

bit and just have those files caught by the `io.k8s.api.*` line.  I think it's useful to just treat core like any of the other `io.k8s.api.*` submodules.

chazsconi commented on lib/kazan/models.ex:
```diff
@@ -52,17 +52,14 @@ defmodule Kazan.Models do
       kind -> {:ok, kind}
       Map.has_key?(data, "kind") and Map.has_key?(data, "apiVersion") ->
-        version = case data["apiVersion"] do
-          "v" <> v ->
```
I changed this because it would fail if any api name started with "v"

obmarg replied:
Ah, good spot.

chazsconi commented on test/models_test.exs:
```diff
@@ -2,13 +2,13 @@ defmodule KazanModelsTest do
   alias Kazan.Models
-  alias Kazan.Models.Api.V1
-  alias Kazan.Models.Apis.Extensions.V1beta1
-  alias Kazan.Models.Apis.Rbac
```
As well as fixing the namespacing, I changed the aliases to not include the version as this makes the code clearer.  e.g. `%Extensions.V1beta1.Deployment` instead of  `%V1beta1.Deployment{}`

chazsconi commented:
For some reason all the tests have passed in Travis now.   Maybe by adding Elixir 1.6.0 it somehow caused Travis to having a longer timeout.

chazsconi commented:
I've now made the changes as per your suggestions and fixed the tests.  It was slightly more complex in the end as changes were also required in `lib/kazan/models.ex`, which I also refactored slightly.

obmarg commented:
> For some reason all the tests have passed in Travis now.    Maybe by adding Elixir 1.6.0 it somehow caused Travis to having a longer  timeout.

I see you merged in master - this might also be because I updated to OTP 19.2.  Though I'm not sure - I did try rebasing this on top of my elixir 1.6 change and it didn't work at the time.  So it's a bit odd.  but lets not complain.

This all looks good now though - thanks a lot for the contribution 👍
